### ISSUE TITLE
Strange regression in lamda parameter type inference in 1.1.405

### ISSUE THREAD

**Issue by tim-becker** [microsoft/pyright#10923]
**Describe the bug**
In v1.1.404, the following code typechecks

```python
def foo():
    for t in [1]:
        pass
    r: set[int] = set()
    a: set[int] = set()
    while r:
        if a:
            s = sorted(a, key=lambda v: v)
            for t in s:
                if -t in r:
                    a = a
```

In particular, the lambda parameter `v` is inferred to be of type `int`.

In v1.1.405, the following errors are detected:
```
error: Argument type is partially unknown
    Argument corresponds to parameter "key" in function "sorted"
    Argument type is "(v: Unknown) -> Unknown" (reportUnknownArgumentType)
error: Type of parameter "v" is unknown (reportUnknownLambdaType)
error: Return type of lambda is unknown (reportUnknownLambdaType)
```

The above code is a mostly minimized repro, based on real code that started failing when upgrading to v1.1.405. Removing any line  in the repo above hides the error.

The issues was confirmed via VSCode + Pylance (with the pyright backend) as well as the cli.

NOTE: When working on the minimized reproducer, I noticed some variants of this code fail on v1.1.404 as well, suggesting the underlying issue exists in both, but some recent change make it trigger more often in v1.1.405.

**Comment by erictraut** [comment]
I think this is because of a bug fix in 1.1.405. The previous behavior was actually incorrect in that the type evaluation results changed based on the order of evaluation. That's why you were seeing the diagnostic only some of the time. With this bug fix in place, the results are deterministic, which is an improvement over the old behavior.

It looks like you have strict type checking enabled because the error in question is not enabled by default in pyright.

You can assist the type inference by providing an explicit type for the target variable.

```python
            s: list[int] = sorted(a, key=lambda v: v)
```

**Comment by tim-becker** [comment]
Thanks for the quick and detailed response.

It makes sense that this "regression" is caused by the determinism fix. And indeed, we use strict type checking.

It still seems like a bug that the lambda parameter type inference fails, right? It's extremely unclear to me what exactly causes the inference to fail; changing almost anything (e.g. `-t` to `t`) makes the inference succeed. There's not a clear source of type ambiguity.

Of course, this is low priority since we can work around it as you describe.
